boot/freeldr/freeldr/arch/uefi/uefimem.c

```diff
@@ -0,0 +1,309 @@
+        {
+            /* Check if the memory is free */
+            if (MemoryMap[i].MemoryType != LoaderFree)
+            {
+                FrLdrBugCheckWithMessage(
+                    MEMORY_INIT_FAILURE,
+                    __FILE__,
+                    __LINE__,
+                    "Failed to reserve memory in the range 0x%Ix - 0x%Ix for %s",
+                    BaseAddress,
+                    Size,
+                    Usage);
+            }
+        }
+    }
+
+    /* Add the memory descriptor */
+    FreeldrDescCount = AddMemoryDescriptor(MemoryMap,
+                                           UNUSED_MAX_DESCRIPTOR_COUNT,
+                                           BasePage,
+                                           PageCount,
+                                           MemoryType);
+}
+
+VOID
+UefiExitBootServices(VOID);
+
+TYPE_OF_MEMORY
+UefiConvertToFreeldrDesc(EFI_MEMORY_TYPE EfiMemoryType)
+{
+    switch (EfiMemoryType)
+    {
+        case EfiReservedMemoryType:
+            return LoaderReserve;
+        case EfiLoaderCode:
+            return LoaderLoadedProgram;
+        case EfiLoaderData:
+            return LoaderLoadedProgram;
+        case EfiBootServicesCode:
+            return LoaderFirmwareTemporary;
+        case EfiBootServicesData:
+            return LoaderFirmwareTemporary;
+        case EfiRuntimeServicesCode:
+            return LoaderFirmwarePermanent;
+        case EfiRuntimeServicesData:
+            return LoaderFirmwarePermanent;
+        case EfiConventionalMemory:
+            return LoaderFree;
+        case EfiUnusableMemory:
+            return LoaderBad;
+        case EfiACPIReclaimMemory:
+            return LoaderFirmwareTemporary;
+        case EfiACPIMemoryNVS:
+            return LoaderReserve;
+        case EfiMemoryMappedIO:
+            return LoaderReserve;
+        case EfiMemoryMappedIOPortSpace:
+            return LoaderReserve;
+        default:
+            break;
+    }
+    return LoaderReserve;
+}
+
+PFREELDR_MEMORY_DESCRIPTOR
+UefiMemGetMemoryMap(ULONG *MemoryMapSize)
+{
+    EFI_LOADED_IMAGE_PROTOCOL *LoadedImage;
+    UINT32 DescriptorVersion;
+    SIZE_T FreeldrMemMapSize;
+    UINTN DescriptorSize;
+    EFI_STATUS Status;
+    UINTN MapSize;
+    UINTN MapKey;
+    UINT32 Index;
+
+    EFI_GUID EfiLoadedImageProtocol = EFI_LOADED_IMAGE_PROTOCOL_GUID;
+    PFREELDR_MEMORY_DESCRIPTOR FreeldrMem = NULL;
+    EFI_MEMORY_DESCRIPTOR* MapEntry = NULL;
+    UINT32 EntryCount = 0;
+    FreeldrDescCount = 0;
+
+    Status = GlobalSystemTable->BootServices->HandleProtocol(GlobalImageHandle,
+                                                             &EfiLoadedImageProtocol,
+                                                             (VOID**)&LoadedImage);
+    if (Status != EFI_SUCCESS)
+    {
+        TRACE("Failed to find LoadedImageHandlewith status: %d", Status);
+        UiMessageBoxCritical("Unable to initialize memory manager.");
```

Comment:
Suggestion of returning failure?
```suggestion
        UiMessageBoxCritical("Unable to initialize memory manager.");
        return NULL;
```